src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,936 @@
+									"will be same with colocate_with table's")));
+		}
+
+		/*shardCount is either 0 or already same with colocateWith table's*/
+		con->shardCount = colocateWithTableShardCount;
+		con->shardCountIsNull = false;
+
+		Var *colocateWithPartKey = DistPartitionKey(colocateWithTableOid);
+
+		if (con->distributionColumn &&
+			colocateWithPartKey->vartype != con->distributionKey->vartype)
+		{
+			ereport(ERROR, (errmsg("cannot colocate with %s and change distribution "
+								   "column to %s because data type of column %s is "
+								   "different then the distribution column of the %s",
+								   con->colocateWith, con->distributionColumn,
+								   con->distributionColumn, con->colocateWith)));
+		}
+		else if (!con->distributionColumn &&
+				 colocateWithPartKey->vartype != con->originalDistributionKey->vartype)
+		{
+			ereport(ERROR, (errmsg("cannot colocate with %s because data type of its "
+								   "distribution column is different than %s",
+								   con->colocateWith, con->relationName)));
+		}
+	}
+	ConvertTable(con);
+}
+
+
+/*
+ * AlterTableSetAccessMethod changes the access method of the given table. It uses
+ * ConvertTable function to create a new table with the access method and move everything
+ * to that table.
+ *
+ * The local and references tables, tables with references, partition tables and foreign
+ * tables are not supported. The function gives errors in these cases.
+ */
+void
+AlterTableSetAccessMethod(TableConversion *con)
+{
+	Relation relation = try_relation_open(con->relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot alter table because no such "
+							   "distributed table exists")));
+	}
+	relation_close(relation, NoLock);
+
+	EnsureTableNotReferencing(con->relationId);
+	EnsureTableNotReferenced(con->relationId);
+	EnsureTableNotForeign(con->relationId);
+
+	if (PartitionedTable(con->relationId))
+	{
+		ereport(ERROR, (errmsg("you cannot alter access method of a partitioned table")));
+	}
+
+	ConvertTable(con);
+}
+
+
+/*
+ * ConvertTable is used for converting a table into a new table with different properties.
+ * The conversion is done by creating a new table, moving everything to the new table and
+ * dropping the old one. So the oid of the table is not preserved.
+ *
+ * The new table will have the same name, columns and rows. It will also have partitions,
+ * views, sequences of the old table. Finally it will have everything created by
+ * GetPostLoadTableCreationCommands function, which include indexes. These will be
+ * re-created during conversion, so their oids are not preserved either (except for
+ * sequences). However, their names are preserved.
+ *
+ * The dropping of old table is done with CASCADE. Anything not mentioned here will
+ * be dropped.
+ */
+void
+ConvertTable(TableConversion *con)
+{
+	if (con->shardCountIsNull)
+	{
+		con->shardCount = con->originalShardCount;
+		con->shardCountIsNull = false;
```

Comment:
The semantics of shardCountIsNull is a little confusing to me. It does not refer to the shardCount because that's an integer, and it does not refer to the original argument, because we change the value here. 

Would it be possible to have something like `newShardCount` and `oldShardCount`, and we react to `newShardCount != oldShardCount`?

Reply:
It is confusing.
I think `shardCount` and `originalShardCount` are `newShardCount` and `oldShardCount`. I couldn't really remove `shardCountIsNull` and rely on the `shardCount`s because of the checks I do.

I'll remove this value change here so the `shardCountIsNull` will mean if the user specified any `shardCount`.